src/Illuminate/Testing/Constraints/HasInDatabase.php

```diff
@@ -29,18 +30,28 @@ class HasInDatabase extends Constraint
      */
     protected $data;
 
+    /**
+     * The model or table that will be used to search for the data.
+     *
+     * @var \Illuminate\Database\Eloquent\Model|string
+     */
+    protected $model;
+
     /**
      * Create a new constraint instance.
      *
      * @param  \Illuminate\Database\Connection  $database
      * @param  array  $data
+     * @param  Model|string  $model
      * @return void
      */
-    public function __construct(Connection $database, array $data)
+    public function __construct(Connection $database, array $data, $model)
```

Comment:
Instead of changing the signature here, would it be better to do something like this in the `assert` methods?
```php
if ($connection instanceof Model && $connection->getKey()) {
    $data = array_merge($data, [$connection->getKeyName() => $connection->getKey()]);
}
```
So that it's just injecting the model's key into the data that it's checking for _if it isn't already set_.

Reply:
That's definitely possible, but would result in duplicate code in both `assertDatabaseHas()` and `assertDatabaseMissing()` methods, as all roads lead to `HasInDatabase` I moved it there.

Reply:
I also don't really love injecting `$model` here. Doesn't feel like this class should have anything to do with Eloquent.

Reply:
Ok, let me refactor this a little.

Reply:
I have left `HasInDatabase` as is. This should be a little more palatable. Thanks for the push in the right direction @samlev